torchtune/modules/transformer.py

```diff
@@ -4,42 +4,80 @@
 # This source code is licensed under the BSD-style license found in the
 # LICENSE file in the root directory of this source tree.
 import copy
-from typing import Optional
+from abc import abstractmethod
+from typing import Optional, Protocol, runtime_checkable
 
 import torch
 from torch import nn, Tensor
 
-from torchtune.modules import CausalSelfAttention, KVCache
+from torchtune.modules import GroupedQueryAttention, KVCache
 
 
-class TransformerDecoderLayer(nn.Module):
+@runtime_checkable
+class TransformerLayer(Protocol):
```

Comment:
is the purpose of this purely for typing? because it seems to be arbitrarily constraining the self attention and cross attention layers to have certain parameters, and you're inheriting from nn.Module anyways

Reply:
Yeah I understand the motivation here, but personally I don't think we should add this. We have a lot of places that use generic `nn.Module` and don't really explicitly define a contract on forward signature. That obviously has its drawbacks, but I think this is a fair bit of extra code complexity for some pretty loose type signature safety improvements. Happy to have a proper discussion on usage of protocols throughout our nn.Module code (though I already tipped my hand as to my opinion here), but I also don't think we should do it as a one-off for one component.

Reply:
I was trying to enforce a consistent signature for layers, but I'll remove this and go in the direction of using **kwargs to absorb unused args.